src/lib/viem.ts

```diff
@@ -0,0 +1,202 @@
+export const walletAddress = writable<Address | undefined>(undefined);
+export const walletClient = writable<WalletClient | undefined>(undefined);
+export const walletBalance = writable<bigint | undefined>(undefined);
+export const sntBalance = writable<bigint | undefined>(undefined);
+export const sntError = writable<string | undefined>(undefined);
+
+// Derived store for formatted balances
+export const formattedBalance = derived(walletBalance, ($balance) => {
+	if ($balance === undefined) return undefined;
+	return Number(formatEther($balance)).toFixed(4);
+});
+
+export const formattedSntBalance = derived(sntBalance, ($balance) => {
+	if ($balance === undefined) return undefined;
+	return Number(formatUnits($balance, SNT_TOKEN.decimals)).toFixed(4);
+});
+
+// Function to fetch ETH balance
+async function fetchBalance(address: Address) {
+	try {
+		console.log('Fetching ETH balance for address:', address);
+		const balance = await publicClient.getBalance({ address });
+		console.log('Received ETH balance:', balance.toString());
+		walletBalance.set(balance);
+	} catch (error) {
+		console.error('Failed to fetch ETH balance:', error);
+		walletBalance.set(undefined);
+	}
+}
+
+// Function to fetch STT balance
+async function fetchSntBalance(address: Address) {
+	try {
+		console.log('Fetching STT balance for address:', address);
+		const balance = await publicClient.readContract({
+			address: SNT_TOKEN.address,
+			abi: CONTRACT_ABI,
+			functionName: 'balanceOf',
+			args: [address]
+		});
+		console.log('Received STT balance:', balance.toString());
+		sntBalance.set(balance);
+		sntError.set(undefined);
+	} catch (error) {
+		console.error('Failed to fetch STT balance:', error);
+		sntBalance.set(undefined);
+		if (error instanceof Error) {
+			sntError.set(error.message);
+		} else {
+			sntError.set('Failed to fetch STT balance');
+		}
+	}
+}
+
+// Function to connect wallet
+export async function connectWallet() {
+	if (!window.ethereum) {
+		throw new Error('MetaMask not installed');
+	}
+
+	const accounts = await window.ethereum.request({
+		method: 'eth_requestAccounts'
+	}) as string[];
+
+	const address = accounts[0] as Address;
+	console.log('Connected to address:', address);
+
+	const client = createWalletClient({
+		chain: sepolia,
+		transport: custom(window.ethereum)
+	});
+
+	walletAddress.set(address);
+	walletClient.set(client);
+
+	// Fetch initial balances
+	await Promise.all([
+		fetchBalance(address),
+		fetchSntBalance(address)
+	]);
+
+	// Set up balance refresh on block
+	const unwatch = publicClient.watchBlocks({
+		onBlock: async () => {
+			const currentAddress = address;
+			if (currentAddress) {
+				await Promise.all([
+					fetchBalance(currentAddress),
+					fetchSntBalance(currentAddress)
+				]);
+			}
+		}
+	});
+
+	// Listen for account changes
+	window.ethereum.on('accountsChanged', async (newAccounts: string[]) => {
+		if (newAccounts.length === 0) {
+			disconnectWallet();
```

Comment:
Could also `return` after `disconnectWallet()` then the `else{}` block isn't needed.

Just bells and whistles..